Allow deprecated non-symbol access to nested `config_for` hashes

paracycle opened the pull request:
### Summary

A change to `config_for` in https://github.com/rails/rails/pull/33815 and https://github.com/rails/rails/pull/33882 has altered the behaviour of the returned object in a breaking manner. 

Before that change, nested hashes returned from `config_for` could be accessed using non-symbol keys. After the change, all keys are recursively symbolized so non-symbol access fails to read the expected values.

This is a breaking change for any app that might be relying on the nested hashes returned from `config_for` calls, and thus should be deprecated before being removed from the codebase.

### Solution

This PR introduces a temporary `NonSymbolAccessDeprecatedHash` class that recursively wraps any nested hashes inside the `OrderedOptions` object returned from `config_for` and issues a deprecation notice when a non-symbol based access is performed.

This way, any apps that are still relying on the ability to access these nested hashes using non-symbol keys will be able to observe the deprecation notices and have time to implement fixes before non-symbol access is removed for good.

### Other Information

Note that the top-level access to the `OrderedOptions` object returned from `config_for` has indifferent access semantics due to the nature of how `OrderedOptions` works. This functionality is the same for `Rails.config` and `secrets`, so is not changed nor is any deprecation notice issued.

The deprecation is only for nested keys below the top-level, since that is the functionality that is potentially broken.

paracycle committed:
Allow deprecated non-symbol access to nested `config_for` hashes

A change to `Rails::Application.config_for` in
https://github.com/rails/rails/pull/33815 and
https://github.com/rails/rails/pull/33882 has altered the behaviour of
the returned object in a breaking manner. Before that change, nested
hashes returned from `config_for` could be accessed using non-symbol keys.
After the change, all keys are recursively symbolized so non-symbol access
fails to read the expected values.

This is a breaking change for any app that might be relying on the
nested hashes returned from `config_for` calls, and thus should be
deprecated before being removed from the codebase.

This commit introduces a temporary `NonSymbolAccessDeprecatedHash` class
that recursively wraps any nested hashes inside the `OrderedOptions`
object returned from `config_for` and issues a deprecation notice when a
non-symbol based access is performed.

This way, apps that are still relying on the ability to access these
nested hashes using non-symbol keys will be able to observe the
deprecation notices and have time to implement changes before non-symbol
access is removed for good.

A CHANGELOG entry is also added to note that non-symbol access to nested
`config_for` hashes is deprecated.

gmcgibbon commented on railties/lib/rails/application.rb:
```diff
@@ -590,5 +590,46 @@ def build_request(env)
         config.app_middleware + super
       end
+
+      class NonSymbolAccessDeprecatedHash < Hash # :nodoc:
```
Would it make more sense to call this `StringAccessDeprecatedHash` and check for string access or are we genuinely concerned with all cases that aren't symbols? String access is all we are testing for.

paracycle replied:
Since the new behaviour restricts usage to symbol keys only, I wanted to make sure that we generate a deprecation notice for anything that is not a symbol. For all intents and purposes, you are correct, we are really trying to detect string access, but I thought it wouldn't hurt to be more general in this case.

I am open to changing the behaviour and renaming the class, if you think otherwise, though.

gmcgibbon replied:
> I am open to changing the behaviour and renaming the class, if you think otherwise, though.

I don't feel strongly about it. Let's get a second opinion.

rafaelfranca replied:
I'm fine with the name and implementation. We can only have strings and symbols keys anyway.

gmcgibbon commented on railties/lib/rails/application.rb:
```diff
@@ -590,5 +590,46 @@ def build_request(env)
+
+        private
+
+          def convert_key(key) # :doc:
```
Why is `# :doc:` here?

paracycle replied:
Sorry, that's a left-over, will remove.

gmcgibbon commented on railties/lib/rails/application.rb:
```diff
@@ -590,5 +590,46 @@ def build_request(env)
+
+          def convert_key(key) # :doc:
+            unless key.kind_of?(Symbol)
+              ActiveSupport::Deprecation.warn(<<~MESSAGE.squish)
```
Nitpick: Since you're doing a heredoc and squish, you can break up the line below with a newline.

paracycle replied:
Sure, will do.

gmcgibbon commented on railties/test/application/configuration_test.rb:
```diff
@@ -1727,6 +1727,7 @@ def index
       app "development"
 
       assert_equal "bar", Rails.application.config.my_custom_config[:foo]
+      assert_equal "bar", Rails.application.config.my_custom_config["foo"]
```
I would rather test deprecated behaviour in deprecation tests only.

paracycle replied:
I have noted in the `Other Information` section of my PR that this is *not* deprecated behaviour.

> Note that the top-level access to the `OrderedOptions` object returned from `config_for` has indifferent access semantics due to the nature of how `OrderedOptions` works.

Since, `my_custom_config` is an instance of `OrderedOptions`, top-level key access is indifferent. Thus, this test is asserting that that behaviour is not changed.

I could put an `assert_not_deprecated` around the string access assertion if it is going to make it more readable and understandable, though.

gmcgibbon replied:
Ah, my mistake. Leaving it as-is if we don't expect this to be deprecated sounds good to me 👍

gmcgibbon reviewed (commented):
This might also be worth mentioning in the changelog as a bugfix for the next release. Can you add a changelog entry?

rafaelfranca commented on railties/lib/rails/application.rb:
```diff
@@ -590,5 +590,46 @@ def build_request(env)
+        def key?(key)
+          super(convert_key(key))
+        end
+
```
Would `slice`, `extract` and `except` and the `!` versions with string keys work? How about `fetch`?

paracycle replied:
They wouldn't work with the current state of the PR, but I can add similar implementation for those as well.

paracycle replied:
Added support for these methods and a test that asserts they all work properly with string keys with deprecations.

paracycle committed:
Fix assertion excpected/actual order

The assertion from the previous PR had the expected and the actual
values in the wrong order, so when a test failed the error message was
confusing.

This commit fixes the problem by switching the order.

paracycle commented:
I think I've addressed all PR comments and squashed my commits, so this PR should be ready to go if there are no more comments.

Additionally, I've add an extra commit to fix the order of expected/actual parameters of some of the assertions added in https://github.com/rails/rails/pull/33882. Since that is logical separate from the main work in this PR, I've opted to leave that as a separate commit, and didn't squash it with my other commits.

Edouard-chin commented on railties/lib/rails/application.rb:
```diff
@@ -590,5 +590,81 @@ def build_request(env)
         config.app_middleware + super
       end
+
+      class NonSymbolAccessDeprecatedHash < Hash # :nodoc:
```
Implementation seems similar to HWA, why not inheriting from it and just override the `convert_key` method ?

paracycle replied:
That's exactly what I tried first and decided against later, for 2 reasons:

1. HWIA stores keys as strings, but we want symbols. So it felt like I was resisting the base class.
2. Even more importantly, `convert_key` is called during both read *and* write, and there is no way to know from which it is called. That would mean we would show deprecation notices as we are interning the YAML parsed hashes and thus showing notices to people who don't even access hashes via string keys. That would be confusing.

Basically HWIA just was not flexible enough for this purpose. But I did take most of the implementation straight from it.

Hope this addresses your concern.

gmcgibbon replied:
I applied this change to your branch.

> 1\. HWIA stores keys as strings, but we want symbols. So it felt like I was resisting the base class.

This seems to work as expected in my tests, and passes the test suite: `Rails.application.config.my_custom_config[:foo] => {:bar=>["baz"] }`

> 2\. Even more importantly, `convert_key` is called during both read _and_ write, and there is no way to know from which it is called. That would mean we would show deprecation notices as we are interning the YAML parsed hashes and thus showing notices to people who don't even access hashes via string keys. That would be confusing.

I'm not seeing this behaviour. Can you explain how this would happen?

We can revert this to not extend from HWIA if need be, but it seems to work as expected AFAICT.

paracycle replied:
So the `[]=` method of HWIA calls `convert_key`  here: https://github.com/rails/rails/blob/master/activesupport/lib/active_support/hash_with_indifferent_access.rb#L94 which would be called ultimately from a source that has nested hashes. If those hashes don't have string keys, we would get a deprecation notice. I guess what's protecting the code from doing that is the fact that you are calling `.to_sym` before passing `key` to `super`, so that writes always get a symbol key.

I guess that solves it more elegantly than I had it. Thanks for the rework and the simplification. 👍
